src/Icicle/Avalanche/Statement/Flatten/Algorithms.hs

```diff
@@ -0,0 +1,747 @@
+-- | When converting from Core to Avalanche, we need to implement
+--   some Core prims with non-trivial Avalanche.
+
+{-# LANGUAGE NoImplicitPrelude   #-}
+{-# LANGUAGE PatternGuards       #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE OverloadedStrings   #-}
+
+module Icicle.Avalanche.Statement.Flatten.Algorithms where
+
+import              Icicle.Avalanche.Statement.Flatten.Base
+
+import              Icicle.Avalanche.Statement.Statement
+import qualified    Icicle.Avalanche.Prim.Flat      as Flat
+import qualified    Icicle.Avalanche.Prim.Compounds as Flat
+
+import qualified    Icicle.Core.Exp.Prim           as Core
+import qualified    Icicle.Common.Exp.Prim.Minimal as Min
+
+import              Icicle.Common.Base
+import              Icicle.Common.Type
+import              Icicle.Common.Exp
+import              Icicle.Common.Fresh
+
+import              Icicle.Internal.Pretty
+
+import              P
+
+import              Data.Hashable                  (Hashable)
+import              Data.String                    (IsString)
+
+
+type FlatX a n = a -> Exp a n Core.Prim -> (Exp a n Flat.Prim -> FlatM a n) -> FlatM a n
+type StmtX a n = Exp a n Flat.Prim -> FlatM a n
+
+--------------------------------------------------------------------------------
+
+initInt :: Name n -> Exp a n p -> Statement a n p -> Statement a n  p
+initInt acc init
+  = InitAccumulator (Accumulator acc IntT init)
+
+readInt :: Name n -> Name n -> Statement a n p -> Statement a n  p
+readInt local acc
+  = Read local acc IntT
+
+initArr :: ValType -> Name n -> Exp a n p -> Statement a n p -> Statement a n p
+initArr t acc init
+  = InitAccumulator (Accumulator acc (ArrayT t) init)
+
+readArr :: ValType -> Name n -> Name n -> Statement a n p -> Statement a n p
+readArr t local acc
+  = Read local acc (ArrayT t)
+
+initBool :: Name n -> Exp a n p -> Statement a n p -> Statement a n  p
+initBool acc init
+  = InitAccumulator (Accumulator acc BoolT init)
+
+readBool :: Name n -> Name n -> Statement a n p -> Statement a n  p
+readBool local acc
+  = Read local acc BoolT
+
+-- Create a let binding with a fresh name
+slet_
+  :: (Hashable n, Monad m)
```

Comment:
hm, damn. I guess I'll just have to live with right-leaning type signatures. Haskell tooling is so bad